### ISSUE TITLE
OpenStack --disk needs to handle cleanup

### ISSUE THREAD

**Issue by keiranmraine** [VertebrateResequencing/wr#46]
If hosts are spawned using the `--disk` option more checking is required before reuse, options include:

 * Automatically delete the work space before a new job is started (use a true tmp location via `mktemp -d -t <prefix>`).  Bad dockers writing files as root will cause this to fail.
 * Automatically teardown the host on completion.
 * Check the available space before reuse.

**Comment by keiranmraine** [comment]
Running off current dev seems to indicate option 2 (teardown), but it still waits for the keepalive.  Can you confirm this is the case?

**Comment by sb10** [comment]
Current dev hasn't (deliberately) changed anything with regard to this issue. A scheduling bug may terminate servers early though (I'm working on it). I expect to implement option 3.

**Comment by sb10** [comment]
Actually, option 3 will be tricky. I have to implement something like option 1 anyway. Would this issue be solved by these 2 new options for the `wr add` json?:

unique_cwd: if set to true, creates a unique directory in cwd, and `cd`s there before running your command.
cleanup:[success|failure|always|never] deletes the unique directory, using sudo if possible

Basically, do your commands only ever create files within the working directory they are run from? Or can a "bad docker" write files anywhere it feels like? If I say "don't use bad dockers like that", can you?

**Comment by keiranmraine** [comment]
It would be good to use `mktemp -d -t <prefix>` and automatically set $TMPDIR and $HOME based on this even with option 3.

e.g.

```bash
TMP_WR=`mktemp -d -t wr_XXXX`
export TMPDIR=$TMP_WR/tmp
export HOME=$TMP_WR/home
mkdir -p $TMPDIR $HOME
```

**Comment by dkj** [comment]
I think setting ```TMPDIR``` per job is a good idea.
Setting that to be the current working directory of the job is _probably_ also good.
Automatic deletion of that ```TMPDIR``` at the end of the job is also _probably_ good.
I'd like ```HOME``` to be constant per instance setup, not per job, as  I setup default iRODS and S3 config (inc. credentials) with the startup script.

**Comment by keiranmraine** [comment]
@djk, that may be fine for your use case, but for the case of mixed use tenants each jobs needs to be separate.  Each job could be reading/writing to different end points or writing as different quotas.  The ENV injection is more appropriate for pushing in keys for RADOS etc.  If you need to add credential files you can inject them at runtime.  For example I use the following to push in the job config, and the ENV json to pass in the S3 keys.

```
echo '{ "reference": { "path": "s3://dockstore-cgp/human/core_ref_GRCh37d5.tar.gz", "class": "File" }, "bwa_idx": { "path": "s3://dockstore-cgp/human/bwa_idx_GRCh37d5.tar.gz", "class": "File" }, "bams_in": [ {"class": "File", "path": "s3://dbGaP-11778/SRR3438345.bam"} ], "cram": false, "sample": "LP6007490_F28green_TumorF28-1", "out_bam": { "path": "/tmp/LP6007490_F28green_TumorF28-1.bam", "class": "File" }, "out_bai": { "path": "/tmp/LP6007490_F28green_TumorF28-1.bam.bai", "class": "File" }, "out_bas": { "path": "/tmp/LP6007490_F28green_TumorF28-1.bam.bas", "class": "File" }, "out_md5": { "path": "/tmp/LP6007490_F28green_TumorF28-1.bam.md5", "class": "File" }, "out_met": { "path": "/tmp/LP6007490_F28green_TumorF28-1.bam.met", "class": "File" }, "out_maptime": { "path": "/tmp/LP6007490_F28green_TumorF28-1.bam.maptime", "class": "File" } } ' > run.json \
&& dockstore tool launch --entry quay.io/wtsicgp/dockstore-cgpmap:1.0.8 --json run.json \
&& s3cmd put --force /tmp/LP6007490_F28green_TumorF28-1* s3://dbGaP-11778/map/ \
&& s3cmd del s3://dbGaP-11778/SRR3438345.bam
```

**Comment by sb10** [comment]
@dkj and @keiranmraine A new proposal; thoughts? This covers not just how cleanup should work, but also how we can make it easy and efficient to deal with files in s3.

    mycmd --ref %wr_input:refs/human.fa%s3://refs/GRCh38.fa* --input %wr_input:input.bam%s3://inputs/foo.bam > %wr_output:output.cram%s3://outputs/bar.cram {"cwd":"/tmp","unique_cwd":true,"unique_home":true,"post_behaviours":[{"when":"always","what":"cleanup"}]}

This would result in the (equivalent of the) following (where uuid is a unique value, not literal):

1. `mkdir /tmp/uuid`
2. `mkdir /tmp/uuid/tmp`
3. `export TMPDIR=/tmp/uuid/tmp`
4. `mkdir /tmp/uuid/home`
5. `export HOME=/tmp/uuid/home` [only if unique_home option was true]
6. `cd /tmp/uuid/home`
7. `mkdir /tmp/wr_imports`
8. [get a list of all files in s3 with prefix GRCh38.fa in refs bucket, and foreach, download to /tmp/wr_imports/refs/ if not already there]
9. `ln /tmp/wr_imports/refs/GRCh38.fa /tmp/uuid/home/refs/human.fa` [repeated for the other files]
10. [download s3://inputs/foo.bam to /tmp/wr_imports/inputs/foo.bam if not already there]
11. `ln /tmp/wr_imports/inputs/foo.bam /tmp/uuid/home/input.bam`
12. `mycmd --ref refs/human.fa --input input.bam > output.cram`
13. [upload output.cram to s3://outputs/bar.cram]
14. `cd /tmp`
15. `sudo rm -fr /tmp/uuid`
16. [wait until the next job resolves steps 1-11, then for each file downloaded to /tmp/wr_imports, check how many hard links there are, and delete if 0]

The purpose of downloading to /tmp/wr_imports and then creating links is to allow multiple jobs running on the same instance with the same inputs to share a download, avoiding repeated downloads.

To address issue #54, if we also included in our `wr add` JSON "volume_cwd":true,"disk":"75G", and changed the cleanup behaviour to only happen when "success", then we instead:

- submit the command with low memory, single core requirements, getting it to run on a cheap flavor
- do step 1 from above
- create and attach a volume 75G large
- mount the volume at /tmp/uuid
- do steps 2-6 from above
- directly download all our inputs from s3 to the desired locations under /tmp/uuid/home
- unmount and detach the volume, recording the volume's uuid against this job
- exit with a special status that says we completed staging
- submit the command again with the proper memory and core requirements, getting it to re-run on the expensive flavor
- do step 1 from above
- mount the recorded volume at /tmp/uuid
- do steps 3, 5 and 6 from above
- do steps 12-14 from above
- unmount and detach the volume
- if step 12 was successful, delete the volume, otherwise leave it and display the volume's identifier in `wr status` against the failed job

There's still an issue of how to handle things like the cram cache, which is too large to waste time actually downloading every time, so we need some way of saying "just fuse mount s3 for this". Other than that, are there any holes here? Are there commands that can't be properly specified and work under this scheme?

For commands that rely on input files that aren't in any way mentioned on the command line, we'd have `wr add` JSON like {"inputs":["extra.input%s3://inputs/extra.input"]}?

**Comment by dkj** [comment]
TL;DR:  not of immediate value to me alas.

The reference counting of input files is interesting. However, my current workloads rarely reuse my main big data inputs, and if they did it would only be for a small set of jobs so WR would also need to juggle such jobs so that they were likely to start on the same instance.

Where I do reuse (much smaller) files is for reference and cram_cache (a case you mention I realise). There I am already relying on a caching fuse layer for the cram_cache as I could not be bothered to work out my exact file requirements, so the actual reference files used by aligners etc effectively come for free with that. The next alternative I'd consider is baking such reference and cram_cache files into the images I create. Then, mounting read-only snapshots volumes of a volume with them on.

I haven't found the access to input data from and dumping output data to S3 too onerous to do within my job scripts so far. (Though I'm still unclear about the reliability and error modes of both standard clients and FUSE layers to do that).

The interpolation of strings in the cmd makes me very slightly nervous (I'd quite like the capability to turn that off if you go this route).

So, alas, I don't think this would add much if any value for me in the short term.

@keiranmraine won't Dockstore give you such inputs & outputs via S3 functionality and are you going to go that route?

**Comment by sb10** [comment]
@dkj What do your "real" commands look like? Do you do something like s3fs mount your bucket at /mnt/bucket, and then your real commands are like `mycmd --ref /mnt/bucket/refs/human.fa --input /mnt/bucket/inputs/input.bam`?

Are you writing straight to the mount as well?

Ignoring the s3 related stuff, does the cleanup and unique_cwd side of things look useful? You'd be able to set those as command line args to `wr add` as well.

**Comment by keiranmraine** [comment]
@sb10, I echo @djk on S3 downloads.  At present all of my work is under dockstore which handles staging of data from pretty much anywhere (ftp/http/S3/synapse, dh3 interested in adding iRODS plugin) and back to various types.

If there are reference files it makes more sense to use a read-only S3 mount (I actually pull a tar.gz from our FTP).

**Comment by sb10** [comment]
@keiranmraine I see; please update issue #54 on if you still need that functionality, and if you do, how would you want it to work with dockstore?

Is the way dockstore gets its data and runs your commands compatible with my unique_cwd and cleanup proposal? I presume you still require the ability to fully clean up before running your next dockstore command on the same instance.

**Comment by dkj** [comment]
I think unique_cwd is a good thing, and associated automatic cleanup also good. So much so that I'd appreciate them on by default.

For 'post processing' options what could be handy is the ability to specify on non-zero exit script - which could be per command but most usefully would be per deploy/manager (as for some of the other cloud options) and would run from the same cwd (pre cleanup) so that comprehensive error logs could be bundled up and dumped somewhere.

Here are two example commands:

1. 

```
export HOME=~ubuntu && echo $HOME && export REF_PATH=/repository/cram_cache/%2s/%2s/%s && mkdir -p 18812_2#4/in 18812_2#4/out 18812_2#4/tmp && cd 18812_2#4 && ( iget -K /seq/18812/18812_2#4.cram in/18812_2#4.cram || (sleep 30; iget -f -K /seq/18812/18812_2#4.cram in/18812_2#4.cram) || (sleep 60; iget -f -K /seq/18812/18812_2#4.cram in/18812_2#4.cram) || (sleep 120; iget -f -K /seq/18812/18812_2#4.cram in/18812_2#4.cram)) && p4dir=`which vtfp.pl` && p4dir=`readlink -f $p4dir` && p4dir=`dirname $p4dir` && p4dir=`dirname $p4dir` && echo p4dir=$p4dir && export PERL5LIB=$p4dir/lib/perl5 && vtfp.pl  -l realign.vtf.log -o realign.json -keys indatadir -vals in -keys outdatadir -vals out -keys  cfgdatadir -vals $p4dir/data/vtlib/ -keys tmpdir -vals tmp -keys rpt -vals 18812_2#4 -keys alignment_method -vals bwa_mem  -keys bwa_executable -values bwa -keys reference_dict -vals /repository/references/Homo_sapiens/GRCh38_full_analysis_set_plus_decoy_hla/all/picard/Homo_sapiens.GRCh38_full_analysis_set_plus_decoy_hla.fa.dict -keys reference_genome_fasta -vals /repository/references/Homo_sapiens/GRCh38_full_analysis_set_plus_decoy_hla/all/fasta/Homo_sapiens.GRCh38_full_analysis_set_plus_decoy_hla.fa -keys alignment_reference_genome -vals /repository/references/Homo_sapiens/GRCh38_full_analysis_set_plus_decoy_hla/all/bwa0_6/Homo_sapiens.GRCh38_full_analysis_set_plus_decoy_hla.fa -keys aligner_numthreads -vals 16 -keys samtools_executable -vals samtools -prune_nodes 'fop.*samtools_stats_F0.*00_bait.*' $p4dir/data/vtlib/realignment_wtsi_template.json && viv.pl -s -v 3 -x -o  viv_realign_1.log ./realign.json && rm out/*.ba{i,m,m.md5} && s3cmd put --multipart-chunk-size-mb 300 --recursive out/ s3://npg_cloud_realign_wip/SC_WES_INT5899948/18812_2#4/ ; es=$?; rm -r in out tmp; exit $es
```
where I pull input data from iRODS and could really do with some arbitrary multiple resource counter mechanism, and push result up to S3 with s3cmd - note how I'm capturing the exit code from analysis and upload, cleaning up, then spitting out that captured error code.

2.
```
{
  "cpus": 4,
  "cloud_flavor": "k1\\.2xlarge",
  "rep_grp": "Interval",
  "cmd": "echo HOME:$HOME PWD:$PWD && export REF_PATH=/repository/cram_cache/%2s/%2s/%s && export NPG_REPOSITORY_ROOT=/repository/ && export PERL5LIB=/software/npg/devel/lib/perl5 && export PATH=/software/npg/devel/bin/:$PATH && which vtfp.pl && L=SC_WES_INT5948441 && ls /npg_cloud_realign_wip/$L && mkdir -p /npg_cloud_realign_wip/$L/merge/ $L && pushd $L && ref=$NPG_REPOSITORY_ROOT/references/Homo_sapiens/GRCh38_full_analysis_set_plus_decoy_hla/all/fasta/Homo_sapiens.GRCh38_full_analysis_set_plus_decoy_hla.fa && vtfp.pl -o tmpout.json  $(echo /npg_cloud_realign_wip/$L/*_*/*.{cram,seqchksum} | perl -ple 's/\\S+sha512primesums512\\S+ ?//g; s/(\\S+cram\\b)/-key incrams -val $1/g; s/(\\S+seqchksum\\b)/-key incrams_seqchksum -val $1/g;' ) -keys cfgdatadir -vals $(dirname $(readlink -f $(which vtfp.pl)))/../data/vtlib/ -keys samtools_executable -vals samtools -keys outdatadir -vals /npg_cloud_realign_wip/$L/merge/  -keys basic_pipeline_params_file -vals $(dirname $(readlink -f $(which vtfp.pl)))/../data/vtlib//alignment_common.json -keys bmd_resetdupflag_val -vals 1  -keys final_output_prep_name -vals final_output_prep -key fopid -val $L -key library -val $L -key br_numthreads_val -val 4 -key b2c_mt_val -val 8 -key b2c_compress_level -val -7jZ -key scramble_reference_fasta -val $ref -prune_nodes '.*(samtools_stats_F0.*00_bait|calibration_pu)-' -splice_nodes bammarkduplicates:-000_bmd_multiway: $(dirname $(readlink -f $(which vtfp.pl)))/../data/vtlib//merge_aligned.json && viv.pl -v 3 -x -s -o viv.log tmpout.json",
  "retries": 0
}
```
where that's all on one line when I pipe it into `wr add`, and both the input and output are on an s3fs, non caching, mount set up in the setup script. (There is no `cloud_flavor` option here right?)

**Comment by keiranmraine** [comment]
@sb10 re: https://github.com/VertebrateResequencing/wr/issues/46#issuecomment-288703435

It would be helpful for WR to allow staging of large files to volume from S3 but only for workloads that are using attached volumes.  With dockstore I can tell it the files are anywhere so if they are local before it starts that's great.  My current use cases would be about 30/70 split for local-SSD vs mounted volume, but if the inputs can be on a mounted volume and the remainder of IO to the local SSD then I can try different I/O patterns.

**Comment by sb10** [comment]
wr add's defaults now implement a solution: the actual working directory is unique, as is TMPDIR, changing HOME is an option, and everything gets deleted after the job completes. The old behaviour can be restored with --cwd_matters.
